API/src/main/java/de/brandwatch/minianalytics/api/controller/MentionController.java

```diff
@@ -2,64 +2,67 @@
-import org.springframework.http.ResponseEntity;
+import org.springframework.format.annotation.DateTimeFormat;
+import org.springframework.http.HttpStatus;
 import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.PathVariable;
 import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.server.ResponseStatusException;
 
-import java.time.Instant;
+import java.time.DateTimeException;
 import java.time.LocalDateTime;
-import java.time.ZoneId;
-import java.time.ZonedDateTime;
-import java.time.format.DateTimeFormatter;
-import java.time.temporal.TemporalAccessor;
+import java.time.ZoneOffset;
 import java.util.List;
 
 @RestController
 public class MentionController {
 
-    private static final Logger logger = LoggerFactory.getLogger(RestController.class);
+    private static final Logger logger = LoggerFactory.getLogger(MentionController.class);
 
     private static final Gson gson = new Gson();
 
-    private final MentionRepository mentionRepository;
+    private final MentionService mentionService;
 
     @Autowired
-    public MentionController(MentionRepository mentionRepository) {
-        this.mentionRepository = mentionRepository;
+    public MentionController(MentionService mentionService) {
+        this.mentionService = mentionService;
     }
 
     @GetMapping(value = "/mentions/{queryID}")
-    public ResponseEntity findMentionsFromQuery(
+    public List<Mention> findMentionsFromQuery(
             @PathVariable String queryID,
-            @RequestParam(value = "date", defaultValue = "") String date){
-
-        logger.info("GET /mentions/" + queryID);
-
-        List<Mention> mentions = mentionRepository.findByQueryID(Long.parseLong(queryID));
-
-        if(!date.equals("")){
-            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
-
-            String convertedDate = date + " 00:00:00";
-            TemporalAccessor temporalAccessor = formatter.parse(convertedDate);
-            LocalDateTime localDateTime = LocalDateTime.from(temporalAccessor);
-            ZonedDateTime zonedDateTime = ZonedDateTime.of(localDateTime, ZoneId.systemDefault());
-            Instant instant = Instant.from(zonedDateTime);
-
-            String dateBounds = "[" + instant + " TO *]";
-
-            logger.info("queryID: *?0* AND date: *?1*");
-
-            return ResponseEntity.status(200).body(gson.toJson(mentionRepository.findMentionsAfterDate(
-                    Long.parseLong(queryID), dateBounds)));
+            @RequestParam(value = "startDate", defaultValue = "")
+            @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime startDate,
+            @RequestParam(value = "endDate", defaultValue = "")
+            @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime endDate) {
+        try {
+            if (startDate == null) {
```

Comment:
I would extract all the date stuff into methods, it's way too distracting in this method. Imagine a method like this:

```java
private Instant parseOrDefault(LocalDateTime date, int minusDays) {
    return date == null ? Instant.now().minus(minusDays, ChronoUnit.DAYS) : date.toInstant(ZoneOffset.UTC);
}
```

that you use like this:

```java
Instant parsedStartDate = parseOrDefault(startDate, 1);
Instant parsedEndDate = parseOrDefault(endDate, 0);
```
and then after that call
```java
validateStartDate(parsedStartDate, parsedEndDate);
```

that does all validation. You can use `Preconditions` for that:

```java
private void validateDates(Instant startDate, Instant endDate) {
    Preconditions.checkArgument(startDate.isBefore(Instant.now()),
            "startDate " + startDate.toString() + " is in the future");
    Preconditions.checkArgument(startDate.isBefore(endDate),
            "startDate " + startDate.toString() + " is after endDate " + endDate.toString());
}
```